lib/ood_core/job/adapters/sge/batch.rb

```diff
@@ -0,0 +1,194 @@
+  require "tempfile"
+  require 'time'
+
+  begin
+    require "ood_core/job/adapters/drmaa"
+    require 'singleton'
+
+    DRMMA_TO_OOD_STATE_MAP = {
+      DRMAA::STATE_UNDETERMINED          => :undetermined,
+      DRMAA::STATE_QUEUED_ACTIVE         => :queued,
+      DRMAA::STATE_SYSTEM_ON_HOLD        => :queued_held,
+      DRMAA::STATE_USER_ON_HOLD          => :queued_held,
+      DRMAA::STATE_USER_SYSTEM_ON_HOLD   => :queued_held,
+      DRMAA::STATE_RUNNING               => :running,
+      DRMAA::STATE_SYSTEM_SUSPENDED      => :suspended,
+      DRMAA::STATE_USER_SUSPENDED        => :suspended,
+      DRMAA::STATE_USER_SYSTEM_SUSPENDED => :suspended,
+      DRMAA::STATE_DONE                  => :completed,
+      DRMAA::STATE_FAILED                => :completed
+    }
+
+    # The one and only connection with DRMAA
+    # Attempting to instantiate a DRMAA::Session more than once causes it to crash
+    class SessionSingleton < DRMAA::Session
+      include Singleton
+    end
+  rescue LoadError; end
+
+  class Error < StandardError; end
+
+  # @param opts [#to_h] the options defining this adapter
+  # @option opts [Batch] :batch The Sge batch object
+  #
+  # @api private
+  # @see Factory.build_sge
+  def initialize(config)
+    @cluster          = config.fetch(:cluster, nil)
+    @conf             = Pathname.new(config.fetch(:conf, nil))
+    @begin            = Pathname.new(config.fetch(:bin, nil))
+    @sge_root         = config.key?(:sge_root) && config[:sge_root] ? Pathname.new(config[:sge_root]) : nil
+
+    @helper = OodCore::Job::Adapters::Sge::Helper.new
+  end
+
+  # Get OodCore::Job::Info for every enqueued job, optionally filtering on owner
+  # @param owner [#to_s] the owner or owner list
+  # @return [Array<OodCore::Job::Info>]
+  def get_all(owner: nil)
+    listener = QstatXmlRListener.new
+    argv = ['qstat', '-r', '-xml']
+    argv += ['-u', owner] unless owner.nil?
+    REXML::Parsers::StreamParser.new(call(*argv), listener).parse
+
+    listener.parsed_jobs.map{|job_hash| OodCore::Job::Info.new(**post_process_qstat_job_hash(job_hash))}
+  end
+
+  # Get OodCore::Job::Info for a job_id that may still be in the queue
+  # 
+  # If @sge_root is nil or libdrmaa is not loaded then we cannot use DRMAA. Using
+  # DRMAA provides better job status and should always be chosen if it is possible.
+  # 
+  # @param job_id [#to_s]
+  # @return [OodCore::Job::Info]
+  def get_info_enqueued_job(job_id)
+    job_info = OodCore::Job::Info.new(id: job_id.to_s, status: :completed)
+    begin
+      listener = QstatXmlJRListener.new
+      argv = ['qstat', '-r', '-xml', '-j', job_id.to_s]
+      REXML::Parsers::StreamParser.new(call(*argv), listener).parse
+
+      job_hash = listener.parsed_job
+      job_hash[:status] = get_status_from_drmma(job_id) if can_use_drmaa?
+        
+      job_info = OodCore::Job::Info.new(**job_hash)
+    rescue Error, REXML::ParseException  # Job not found
+    end
```

Comment:
Why rescue Error here? In this case `call(*argv)` can result in an error, if for example, the scheduler is unavailable, or OnDemand has been installed on a new node and that node hasn't been added to the list of hosts SGE will accept qstat calls from, etc.

In the error case I would expect the error to bubble up. Instead it will be caught, ignored, and the job will be reported as completed.

Also, would a REXML::ParseException occur if the output of the `QstatXmlJRListener` doesn't handle the output properly say in the case that the XML is missing a tag or the format is different? Or is the REXML::ParseException only to ever occur if the xml is well formatted but the job id is not returned by the call?

Oh perhaps when the job id doesn't exist the qstat call returns an exit status not 0?

Reply:
I have removed the catch on `Error`. As for `ParseException` when `qstat -xml -j` fails to find a job it responds with invalid XML...because design? I confirmed this confusing behavior on both my development VM and UCLA's Hoffman cluster so it appears reliable.

Reply:
I guess one final risk here is the assumption that results will contain a newline :-)

```
results.split("\n")[1]
```

What about just a regex:

```ruby
raise e unless results =~ /unknown_jobs/
```

This actually also works if results happened to be nil for whatever reason.

Reply:
Is it worth customizing the message, something like:

```ruby
raise e, "REXML::ParseException error and command '#{argv.join(' ')' produced results that didn't contain string 'unknown_jobs'. ParseException: #{e.message}"
```

Reply:
I think that is probably the safest thing to do.